Remove #![allow(rustc::potential_query_instability)] from rustc_infer

CastilloDel opened the pull request:
Related to #84447

This PR probably needs to be benchmarked to check for regressions.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @jackh726 (or someone else) soon.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

CastilloDel commented:
@bors try

bors commented:
@CastilloDel: :key: Insufficient privileges: not in try users

jackh726 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 540d4bb99d79ad1587b7c021a3f99ef9178aa538 with merge 8b67abcd612ebae784d854d3a5c6ad4f428022b1...
<!-- homu: {"type":"TryBuildStarted","head_sha":"540d4bb99d79ad1587b7c021a3f99ef9178aa538","merge_sha":"8b67abcd612ebae784d854d3a5c6ad4f428022b1"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3277572555/jobs/5396029257)
Build commit: 8b67abcd612ebae784d854d3a5c6ad4f428022b1 (`8b67abcd612ebae784d854d3a5c6ad4f428022b1`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3277572555/jobs/5396029257"},"merge_sha":"8b67abcd612ebae784d854d3a5c6ad4f428022b1"} -->

rust-timer commented:
Queued 8b67abcd612ebae784d854d3a5c6ad4f428022b1 with parent a24a020e6d926dffe6b472fc647978f92269504e, future [comparison URL](https://perf.rust-lang.org/compare.html?start=a24a020e6d926dffe6b472fc647978f92269504e&end=8b67abcd612ebae784d854d3a5c6ad4f428022b1).

rust-timer commented:
Finished benchmarking commit (8b67abcd612ebae784d854d3a5c6ad4f428022b1): [comparison URL](https://perf.rust-lang.org/compare.html?start=a24a020e6d926dffe6b472fc647978f92269504e&end=8b67abcd612ebae784d854d3a5c6ad4f428022b1&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=a24a020e6d926dffe6b472fc647978f92269504e&end=8b67abcd612ebae784d854d3a5c6ad4f428022b1&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | 0.2%     | [0.2%, 0.3%] | 9         |
| Regressions ❌ <br /> (secondary) | 1.1%     | [0.6%, 1.6%] | 8         |
| Improvements ✅ <br /> (primary) | -0.5%    | [-0.7%, -0.5%] | 6         |
| Improvements ✅ <br /> (secondary) | -        | -     | 0         |
| All ❌✅ (primary) | -0.1%    | [-0.7%, 0.3%] | 15        |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=a24a020e6d926dffe6b472fc647978f92269504e&end=8b67abcd612ebae784d854d3a5c6ad4f428022b1&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | -        | -     | 0         |
| Regressions ❌ <br /> (secondary) | 2.8%     | [2.2%, 3.2%] | 3         |
| Improvements ✅ <br /> (primary) | -        | -     | 0         |
| Improvements ✅ <br /> (secondary) | -        | -     | 0         |
| All ❌✅ (primary) | -        | -     | 0         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=a24a020e6d926dffe6b472fc647978f92269504e&end=8b67abcd612ebae784d854d3a5c6ad4f428022b1&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | -        | -     | 0         |
| Regressions ❌ <br /> (secondary) | 3.0%     | [3.0%, 3.0%] | 1         |
| Improvements ✅ <br /> (primary) | -        | -     | 0         |
| Improvements ✅ <br /> (secondary) | -3.2%    | [-3.2%, -3.2%] | 1         |
| All ❌✅ (primary) | -        | -     | 0         |
</details>

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

CastilloDel commented:
The result seems okayish to me. In instructions count, it seems the primary improvements overweight the primary regressions. The regressions in secondary benchmarks, specially the memory ones, don't seem to be that good, but I guess they are acceptable?

jackh726 reviewed (approved):
LGTM, just needs rebase

CastilloDel committed:
Remove #![allow(rustc::potential_query_instability)] from rustc_infer

Change reported_violations to use IndexSet

It is being used to iterate and to insert, without a lot of lookups
so hopefully it won't be a perf hit

Change MiniGraph.nodes to use IndexSet

It is being used to iterate and to insert, without performing lookups
so hopefully it won't be a perf hit

Change RegionConstraintData.givens to a FxIndexSet

This might result in a perf hit. Remove was being used in `givens`,
and `FxIndexSet` doesn't allow calling remove without losing the
fixed iteration order. So it was necessary to change remove to
`shift_remove`, but this method is slower.

Change OpaqueTypesVisitor to use stable sets and maps

This could also be a perf hit.

Make TraitObject visitor use a stable set

CastilloDel commented:
Sorry, @/rust-lang/miri I messed up the rebase. It should be okay now!

oli-obk commented:
@bors r=jackh726

bors commented:
:pushpin: Commit efb4a3ea7eb08f0dadb4dbe453350b784b7a9c56 has been approved by `jackh726`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jackh726 efb4a3ea7eb08f0dadb4dbe453350b784b7a9c56 -->
<!-- homu: {"type":"Approved","sha":"efb4a3ea7eb08f0dadb4dbe453350b784b7a9c56","approver":"jackh726","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

CastilloDel commented:
I have changed `shift_remove` to `remove` following the conversation in https://github.com/rust-lang/rust/pull/102674. It should be better for the performance and still work, because we only care about the reproducibility and not about having an certain order.

bors commented:
:umbrella: The latest upstream changes (presumably #103671) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

CastilloDel commented:
Not really sure about what happened, but it should be okay now

jackh726 commented:
@bors r+

bors commented:
:pushpin: Commit e9502010b470fdd60e90e9a2216c2a568882ae30 has been approved by `jackh726`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jackh726 e9502010b470fdd60e90e9a2216c2a568882ae30 -->
<!-- homu: {"type":"Approved","sha":"e9502010b470fdd60e90e9a2216c2a568882ae30","approver":"jackh726","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit e9502010b470fdd60e90e9a2216c2a568882ae30 with merge ca08a3265511d3b0c2f3c3cf06e7acc10b73a5ce...
<!-- homu: {"type":"BuildStarted","head_sha":"e9502010b470fdd60e90e9a2216c2a568882ae30","merge_sha":"ca08a3265511d3b0c2f3c3cf06e7acc10b73a5ce"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3408412011/jobs/5671511893)
Approved by: jackh726
Pushing ca08a3265511d3b0c2f3c3cf06e7acc10b73a5ce to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jackh726","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3408412011/jobs/5671511893"},"merge_sha":"ca08a3265511d3b0c2f3c3cf06e7acc10b73a5ce"} -->

rust-timer commented:
Finished benchmarking commit (ca08a3265511d3b0c2f3c3cf06e7acc10b73a5ce): [comparison URL](https://perf.rust-lang.org/compare.html?start=9b735a7132acd58b3bd34c084e9ca5b4ca7450a2&end=ca08a3265511d3b0c2f3c3cf06e7acc10b73a5ce&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=9b735a7132acd58b3bd34c084e9ca5b4ca7450a2&end=ca08a3265511d3b0c2f3c3cf06e7acc10b73a5ce&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3% | [0.3%, 0.3%] | 1     |
| Regressions ❌ <br /> (secondary)  | 0.5% | [0.5%, 0.5%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.3% | [0.3%, 0.3%] | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=9b735a7132acd58b3bd34c084e9ca5b4ca7450a2&end=ca08a3265511d3b0c2f3c3cf06e7acc10b73a5ce&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.2%  | [0.5%, 1.8%]   | 5     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.5% | [-0.5%, -0.5%] | 1     |
| Improvements ✅ <br /> (secondary) | -4.5% | [-4.5%, -4.5%] | 1     |
| All ❌✅ (primary)                 | 0.9%  | [-0.5%, 1.8%]  | 6     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=9b735a7132acd58b3bd34c084e9ca5b4ca7450a2&end=ca08a3265511d3b0c2f3c3cf06e7acc10b73a5ce&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.8% | [-2.8%, -2.8%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>
